File: src/typescript/vitest-example/example.test.ts
import { describe, test, expect, vi } from "vitest";
import { User, UserService } from "./model";

describe("User creation", () => {
  test("user display name is formatted correctly", () => {
    const user = new User("Charles", "Chen", "chrlschn@example.org");

    expect(user.displayName).toBe("Charles Chen");
  });

  test("user handle is email with @", () => {
    const user = new User("Ada", "Lovelace", "alove@example.org");

    expect(user.displayName).toBe("Ada Lovelace");
    expect(user.handle).toBe("@alove");
  });

  test("user saved to database", () => {
    const UserRepository = vi.fn();
    UserRepository.prototype.saveToDb = vi.fn(
      () => "Saved user (from mock repository)"
    );

    const repo = new UserRepository();
    const userService = new UserService(repo);
    const user = new User("Ada", "Lovelace", "alove@example.org");
    const msg = userService.saveUser(user);

    expect(msg).toBe("Saved user (from mock repository)");
  });
});

File: src/csharp/tunit-example/Example.Test.cs
using NSubstitute;

public class User_Creation
{
  [Test]
  public async Task User_Display_Name_Is_Formatted_Correctly()
  {
    var user = new User("Charles", "Chen", "chrlschn@example.org");

    await Assert.That(user.DisplayName).IsEqualTo("Charles Chen");
  }

  [Test]
  public async Task User_Handle_Is_Email_With_At()
  {
    var user = new User("Ada", "Lovelace", "alove@example.org");

    await Assert.That(user.DisplayName).IsEqualTo("Ada Lovelace");
    await Assert.That(user.Handle).IsEqualTo("@alove");
  }

  [Test]
  public async Task User_Save_To_Database()
  {
    var user = new User("Ada", "Lovelace", "alove@example.org");

    var mockRepo = Substitute.For<UserRepository>();
    mockRepo.SaveToDb(user).Returns("Saved user (from mock repository)");

    var userService = new UserService(mockRepo);
    var msg = userService.SaveUser(user);

    await Assert.That(msg).IsEqualTo("Saved user (from mock repository)");
  }
}
